qutip/core/data/_creator_utils.py

```diff
@@ -24,4 +24,8 @@ def true(arg):
 
 
 def data_copy(arg, shape, copy=True):
+    if shape and shape != arg.shape:
```

Comment:
Probably `shape is not None` to help catch cases of a user doing something super dumb like `shape=()` or `shape=0` by accident.
```suggestion
    if shape is not None and shape != arg.shape:
```